contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -1,436 +1,626 @@
-compatible scheduler in Kubernetes where tens of thousands of users depend on
-the behavior of the scheduler.
-After revisiting the ideas and challenges, we changed our plan and have decided
-to build some of the ideas of the scheduling framework into Scheduler V1 to make
-it more extendable.
-
-As the first step, we would like to build:
- 1. [Pre-bind](#pre-bind) and [Reserve](#reserve)  plugin points. These will
- help us move our existing cluster resource binding code, such as persistent
- volume binding, to plugins.
- 1. We will also build
- [the plugin communication mechanism](#communication-and-statefulness-of-plugins).
- This will allow us to build more sophisticated plugins that would require
- communication and also help us clean up existing scheduler's code by removing
- existing transient cache data.
- 
-More features of the framework can be added to the Scheduler in the future based
-on the requirements.
-
-<s>
-# CONFIGURING THE SCHEDULING FRAMEWORK
-
-TBD
-
-# BACKWARD COMPATIBILITY WITH SCHEDULER v1
-
-We will build a new set of plugins for scheduler v2 to ensure that the existing
-behavior of scheduler v1 in placing Pods on nodes is preserved. This includes
-building plugins that replicate default predicate and priority functions of
-scheduler v1 and its binding mechanism, but scheduler extenders built for
-scheduler v1 won't be compatible with scheduler v2. Also, predicate and priority
-functions which are not enabled by default (such as service affinity) are not
-guaranteed to exist in scheduler v2.
+### Un-reserve
 
-# DEVELOPMENT PLAN
+This is an informational extension point. If a pod was reserved and then
+rejected in a later phase, then un-reserve plugins will be notified. Un-reserve
+plugins should clean up state associated with the reserved Pod.
+
+Plugins that use this extension point usually should also use
+[Reserve](#reserve).
+
+## Plugin API
+
+There are two steps to the plugin API. First, plugins must register and get
+configured, then they use the extension point interfaces. Extension point
+interfaces have the following form.
+
+```go
+type Plugin interface {
+   Name() string
+}
+
+type QueueSortPlugin interface {
+   Plugin
+   Less(*v1.Pod, *v1.Pod) bool
+}
+
+type PreFilterPlugin interface {
+   Plugin
+   PreFilter(PluginContext, *v1.Pod) error
+}
+
+// ...
+```
+
+### PluginContext
+
+Most* plugin functions will be called with a `PluginContext` argument. A
+`PluginContext` represents the current scheduling cycle.
 
-We will develop the scheduling framework as an incubator project in SIG
-scheduling. It will be built in a separate code-base independently from
-scheduler v1, but we will probably use a lot of code from scheduler v1. 
+A `PluginContext` provides read-only APIs for accessing the scheduler's cache of
+cluster state. This is the preferred way for plugins to iterate over nodes,
+iterate over pods on one node, check available resources, and other tasks. The
```

Comment:
Autoscaling cannot rely on cache. It runs predicates in different scenarios, adding or removing nodes, moving pods around, etc. It is not efficient to build scheduler cache from scratch every time CA needs to run some predicates. CA also needs to be able to inject non-existing in-memory objects (mostly nodes, but also other things like PVs for LocalPersistentVolumes) into predicates or remove/modify existing objects for the purpose of running predicates.

I think `PluginContext` those requirements can be addressed by making `PluginContext` an interface that CA can implement itself.